### PR TITLE
nimble/host: Change host-stop procedure

### PR DESCRIPTION
**Pull request by rymanluk**
With this PR host is sending in the raw Disconnect Command to all the active connections and then start graceful disconnect timeout 2s.

Host notifies that is stopped when
a) all the disconnection complete events arrived within 2s
b) disconnect timeout fires

btmon logs for a) scenario:

< HCI Command: Disconnect (0x01|0x0006) plen 3                                                                         #1191 110.578000
        Handle: 3
        Reason: Remote User Terminated Connection (0x13)
> HCI Event: Command Status (0x0f) plen 4                                                                              #1192 110.578000
      Disconnect (0x01|0x0006) ncmd 1
        Status: Success (0x00)
< HCI Command: Disconnect (0x01|0x0006) plen 3                                                                         #1193 110.601500
        Handle: 2
        Reason: Remote User Terminated Connection (0x13)
> HCI Event: Command Status (0x0f) plen 4                                                                              #1194 110.601500
      Disconnect (0x01|0x0006) ncmd 1
        Status: Success (0x00)
< HCI Command: Disconnect (0x01|0x0006) plen 3                                                                         #1195 110.624900
        Handle: 1
        Reason: Remote User Terminated Connection (0x13)
> HCI Event: Command Status (0x0f) plen 4                                                                              #1196 110.624900
      Disconnect (0x01|0x0006) ncmd 1
        Status: Success (0x00)
> HCI Event: Number of Completed Packets (0x13) plen 5                                                                 #1197 110.632700
        Num handles: 1
        Handle: 2
        Count: 0
> HCI Event: Disconnect Complete (0x05) plen 4                                                                         #1198 110.632700
        Status: Success (0x00)
        Handle: 3
        Reason: Connection Terminated By Local Host (0x16)

> HCI Event: Disconnect Complete (0x05) plen 4                                                                         #1201 110.695200
        Status: Success (0x00)
        Handle: 1
        Reason: Connection Terminated By Local Host (0x16)
> HCI Event: Disconnect Complete (0x05) plen 4                                                                         #1202 110.765500
        Status: Success (0x00)
        Handle: 2
        Reason: Connection Terminated By Local Host (0x16)

### CODE REVIEW

File: nimble/host/src/ble_hs_stop.c [hunk 1]
```diff
@@ -33,6 +36,11 @@ static struct ble_gap_event_listener ble_hs_stop_gap_listener;
 SLIST_HEAD(ble_hs_stop_listener_slist, ble_hs_stop_listener);
 static struct ble_hs_stop_listener_slist ble_hs_stop_listeners;
 
+/* Track number of connections */
+static uint8_t conn_cnt;
+
+static struct ble_npl_callout terminate_tmo;
+
```
Comment (ccollins476ad) [inline nimble/host/src/ble_hs_stop.c]: I think these two globals (`conn_cnt`, `terminate_tmo`) should have the `ble_hs_stop_` prefix that the other symbols have.

File: nimble/host/src/ble_hs_stop.c [hunk 2]
```diff
@@ -21,9 +21,12 @@
...
 #include "syscfg/syscfg.h"
 #include "ble_hs_priv.h"
+#include "nimble/nimble_npl.h"
+#ifndef MYNEWT
+#include "nimble/nimble_port.h"
+#endif
 
-static ble_npl_event_fn ble_hs_stop_term_event_cb;
-static struct ble_npl_event ble_hs_stop_term_ev;
+#define BLE_HOST_STOP_TIMEOUT_MS (2000)
```
Comment (ccollins476ad) [inline nimble/host/src/ble_hs_stop.c]: It would be better if this timeout were were configurable.

File: nimble/host/src/ble_gap.c [hunk 3]
```diff
@@ -4469,44 +4469,51 @@ ble_gap_conn_active(void)
...
-    if (conn == NULL) {
-        rc = BLE_HS_ENOTCONN;
-        goto done;
-    }
-
+    BLE_HS_DBG_ASSERT(ble_hs_locked_by_cur_task());
     if (conn->bhc_flags & BLE_HS_CONN_F_TERMINATING) {
-        rc = BLE_HS_EALREADY;
-        goto done;
+        return  BLE_HS_EALREADY;
```
Comment (ccollins476ad) [inline nimble/host/src/ble_gap.c]: nitpick: there's an extra space.

File: nimble/host/src/ble_hs_stop.c [hunk 4]
```diff
@@ -95,47 +105,41 @@ ble_hs_stop_terminate_all_periodic_sync(void)
...
     if (rc == 0) {
         /* Terminate procedure successfully initiated.  Let the GAP event
          * handler deal with the result.
          */
-    } else {
-        BLE_HS_LOG(ERROR,
-            "ble_hs_stop: failed to terminate connection; rc=%d\n", rc);
-        ble_hs_stop_done(rc);
+        conn_cnt++;
+    }  else {
```
Comment (ccollins476ad) [inline nimble/host/src/ble_hs_stop.c]: nitpick: there's an extra space.

File: nimble/host/src/ble_hs_conn.c [hunk 5]
```diff
@@ -199,6 +199,20 @@ ble_hs_conn_delete_chan(struct ble_hs_conn *conn, struct ble_l2cap_chan *chan)
     ble_l2cap_chan_free(chan);
 }
 
+void
+ble_hs_conn_foreach(ble_hs_conn_foreach_fn *cb)
```
[the remark this answers is not in the document]
Comment (rymanluk) [reply nimble/host/src/ble_hs_conn.c]: agreed

File: nimble/host/src/ble_hs_stop.c (hunk 1 above)
Comment (rymanluk) [reply nimble/host/src/ble_hs_stop.c]: of course. thx

File: nimble/host/src/ble_hs_stop.c (hunk 2 above)
Comment (rymanluk) [reply nimble/host/src/ble_hs_stop.c]: agree

**Comment by rymanluk** [comment]
@ccollins476ad Thanks, fixed

**Pull request merged**